src/symbol.rs

```diff
@@ -42,21 +43,48 @@ impl Clone for SymbolText {
         match self {
             SymbolText::Owned(text) => SymbolText::Owned(text.to_owned()),
             SymbolText::Shared(text) => SymbolText::Shared(Rc::clone(text)),
+            SymbolText::Unknown => SymbolText::Unknown,
         }
     }
 }
 
-/// The text of a fully resolved field name, annotation, or symbol value. The text stored in this
-/// Symbol may be either a `String` or a shared reference to text in a symbol table.
-#[derive(Debug, Hash, Clone, Eq)]
+impl PartialEq<Self> for SymbolText {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl PartialOrd<Self> for SymbolText {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for SymbolText {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self.text(), other.text()) {
+            // If both Symbols have known text, delegate the comparison to their text.
+            (Some(s1), Some(s2)) => s1.cmp(s2),
+            // Otherwise, $0 (unknown text) is treated as 'less than' known text
+            (Some(_), None) => Ordering::Greater,
+            (None, Some(_)) => Ordering::Less,
+            (None, None) => Ordering::Equal,
+        }
+    }
+}
+
+/// The text of a fully resolved field name, annotation, or symbol value. If the symbol has known
+/// text (that is: the symbol is not `$0`), it will be stored as either a `String` or a shared
+/// reference to text in a symbol table.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
 pub struct Symbol {
```

Comment:
A `Symbol` represents a _resolved_ symbol, which is to say that this part: 

>  SymbolTokens with undefined text do not necessarily refer to symbols with unknown text because the text may be resolvable through the SymbolToken’s importLocation

has already been attempted. If it failed, then an error would have been returned by the reader and we wouldn't be constructing a `Symbol`. If we have a `Symbol`, then the resolution step has already succeeded.

Reply:
If, after resolution, the `Symbol` is found to explicitly have undefined text (`$0` is the main example), then the `Symbol`'s inner `SymbolText` enum will be set to `Unknown`.